libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/route/line/api/MapboxRouteLineApi.kt

```diff
@@ -794,77 +816,99 @@ class MapboxRouteLineApi(
         routeProgress: RouteProgress,
         consumer: MapboxNavigationConsumer<Expected<RouteLineError, RouteLineUpdateValue>>
     ) {
-        val currentPrimaryRoute = primaryRoute
-        if (currentPrimaryRoute == null) {
-            val msg = "You're calling #updateWithRouteProgress without any routes being set."
-            consumer.accept(
-                ExpectedFactory.createError(RouteLineError(msg, throwable = null))
-            )
-            logW(msg, LOG_CATEGORY)
-            return
-        } else if (currentPrimaryRoute.id != routeProgress.navigationRoute.id) {
-            val msg = "Provided primary route (#setNavigationRoutes, ID: " +
-                "${currentPrimaryRoute.id}) and navigated route (#updateWithRouteProgress, ID: " +
-                "${routeProgress.navigationRoute.id}) are not the same. Aborting the update."
-            consumer.accept(
-                ExpectedFactory.createError(RouteLineError(msg, throwable = null))
-            )
-            logE(msg, LOG_CATEGORY)
-            return
-        }
+        routeProgressJobControl.job.cancelChildren()
```

Comment:
This can potentially starve the queue, can't it? If the progress updates come at a faster rate then we can process them, then we run into a risk of continuously canceling the processing job and never finishing.

We should only cancel a queued job if it exists.

Reply:
The idea here is that the only suspend method in `routeProgressJobControl` scope is `mutex.lock()`. So, if we're waiting for lock, it means the job is queued. If we are not, the update is in progress. But `cancelChildren()` will only cancel the `mutex.lock()`. Once we start a new job inside using `jobControl` (not `routeProgressJobControl`), it won't get cancelled. And everything else is not suspend.
Does this make sense? Too complicated?
I was thinking about writing my own implementation of a coalescing queue, but I remember that when last time I tried to do it, we discussed with @VysotskiVadim and decided this functionality is already supported by coroutines, why not just use them.

Reply:
While this makes sense and works, we are only one suspension away from breaking this pattern. We need to make this extremely clear that the contents of the task cannot have any suspensions. Perhaps when we have the bandwidth we should revisit this setup to make it more robust and allow for nested suspensions.